File: sdk/workbuddy-enterprises/python-sdk/src/workbuddy_enterprise/resources/members.py
from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any

from workbuddy_enterprise._serialization import clean_dict
from workbuddy_enterprise.errors import WorkBuddyConfigError
from workbuddy_enterprise.pagination import page_query
from workbuddy_enterprise.resources._base import Resource
from workbuddy_enterprise.response import ApiResponse, Page


class MembersResource(Resource):
    def list(
        self,
        *,
        page_num: int | None = None,
        page_size: int | None = None,
        keyword: str | None = None,
    ) -> ApiResponse[Page[dict[str, Any]]]:
        params = {"keyword": keyword, **page_query(page_num=page_num, page_size=page_size)}
        return self._as_page(self._get("/openapi/members", params=params))

    def add(
        self,
        members: Sequence[Mapping[str, Any]],
        *,
        grant_license: bool | None = None,
        **body_fields: Any,
    ) -> ApiResponse[dict[str, Any]]:
        """Batch add members.

        YAML requires body.members[] with at least username + email.
        """
        member_payloads = [dict(member) for member in members]
        for index, member in enumerate(member_payloads):
            missing = [key for key in ("username", "email") if key not in member]
            if missing:
                raise WorkBuddyConfigError(
                    f"members.add members[{index}] requires: {', '.join(missing)}",
                )
        payload: dict[str, Any] = {"members": member_payloads}
        if grant_license is not None:
            payload["grantLicense"] = grant_license
        payload.update(clean_dict(body_fields))
        return self._as_map(self._post_json("/openapi/members/add", body=payload))

File: sdk/workbuddy-enterprises/rust-sdk/src/resources/members.rs
use crate::client::{push_q, push_qi, Client};
use crate::error::Result;
use crate::operations::validate_json_operation;
use crate::response::ApiResponse;
use serde_json::{json, Value};

pub struct MembersResource<'a> {
    pub(crate) client: &'a Client,
}

impl MembersResource<'_> {
    pub fn list(
        &self,
        page_num: Option<i64>,
        page_size: Option<i64>,
        keyword: Option<&str>,
    ) -> Result<ApiResponse<crate::response::Page<Value>>> {
        let mut q = Vec::new();
        push_qi(&mut q, "pageNum", page_num);
        push_qi(&mut q, "pageSize", page_size);
        push_q(&mut q, "keyword", keyword.map(|s| s.to_string()));
        self.client.get_page("/openapi/members", &q)
    }

    /// YAML requires body.members[] with username + email.
    pub fn add(
        &self,
        members: &[Value],
        grant_license: Option<bool>,
    ) -> Result<ApiResponse<Value>> {
        let mut body = serde_json::Map::new();
        body.insert("members".into(), json!(members));
        if let Some(g) = grant_license {
            body.insert("grantLicense".into(), json!(g));
        }
        let body = Value::Object(body);
        validate_json_operation("members-add", &body)?;
        self.client.post_json("/openapi/members/add", &[], body)
    }

    pub fn add_raw(&self, body: Value) -> Result<ApiResponse<Value>> {
        validate_json_operation("members-add", &body)?;
        self.client.post_json("/openapi/members/add", &[], body)
    }
}
